stan/math/rev/core/arena_matrix.hpp

```diff
@@ -139,6 +135,264 @@ class arena_matrix : public Eigen::Map<MatrixType> {
+   * @tparam ArenaMatrix An `arena_matrix` type
+   * @param other matrix to copy from
+   * @return `*this`
+   */
+  template <typename ArenaMatrix,
+            require_same_t<ArenaMatrix, arena_matrix<MatrixType>>* = nullptr>
+  arena_matrix& operator=(ArenaMatrix&& other) {
+    // placement new changes what data map points to - there is no allocation
+    new (this) Base(other.rows(), other.cols(), other.nonZeros(),
+                    const_cast<StorageIndex*>(other.outerIndexPtr()),
+                    const_cast<StorageIndex*>(other.innerIndexPtr()),
+                    const_cast<Scalar*>(other.valuePtr()));
+    return *this;
+  }
+
+  /**
+   * Assignment operator for assigning an expression.
+   * @tparam Expr An expression that an `arena_matrix<MatrixType>` can be
+   * constructed from
+   * @param expr expression to evaluate into this
+   * @return `*this`
+   */
+  template <typename Expr,
+            require_not_same_t<Expr, arena_matrix<MatrixType>>* = nullptr>
+  arena_matrix& operator=(Expr&& expr) {
+    *this = arena_matrix(std::forward<Expr>(expr));
+    return *this;
+  }
+
+ private:
+  /**
+   * inplace operations functor for `Sparse (.*)= Sparse`.
+   * @note This assumes that each matrix is of the same size and sparsity
+   * pattern.
+   * @tparam F A type with a valid `operator()(Scalar& x, const Scalar& y)`
+   * method
+   * @tparam Expr Type derived from `Eigen::SparseMatrixBase`
+   * @param f Functor that performs the inplace operation
+   * @param other The right hand side of the inplace operation
+   */
+  template <typename F, typename Expr,
+            require_convertible_t<Expr&, MatrixType>* = nullptr>
+  void inplace_ops_impl(F&& f, Expr&& other) {
+    auto&& x = to_ref(other);
+    for (int k = 0; k < (*this).outerSize(); ++k) {
+      typename Base::InnerIterator it(*this, k);
+      typename std::decay_t<Expr>::InnerIterator iz(x, k);
+      for (; static_cast<bool>(it) && static_cast<bool>(iz); ++it, ++iz) {
+        f(it.valueRef(), iz.value());
+      }
+    }
+  }
+
+  /**
+   * inplace operations functor for `Sparse (.*)= Dense`.
+   * @note This assumes the user intends to perform the inplace operation for
+   * the nonzero parts of `this`
+   * @tparam F A type with a valid `operator()(Scalar& x, const Scalar& y)`
+   * method
+   * @tparam Expr Type derived from `Eigen::DenseBase`
+   * @param f Functor that performs the inplace operation
+   * @param other The right hand side of the inplace operation
+   */
+  template <typename F, typename Expr,
+            require_not_convertible_t<Expr&, MatrixType>* = nullptr,
+            require_eigen_dense_base_t<Expr>* = nullptr>
+  void inplace_ops_impl(F&& f, Expr&& other) {
+    auto&& x = to_ref(other);
+    for (int k = 0; k < (*this).outerSize(); ++k) {
+      typename Base::InnerIterator it(*this, k);
+      for (; static_cast<bool>(it); ++it) {
+        f(it.valueRef(), x(it.row(), it.col()));
+      }
+    }
```

Comment:
Are these operations something that could be implemented using Eigen's `NullaryExpr` framework? That could allow more room for Eigen to optimise/simplify ops

Reply:
I think `NullaryExpr` is only available for dense types :/

If we wanted to optimize this harder we could write specific `operator+=` for each of sparse, dense, and scalar ops that under the hood will make an Eigen map to a dense eigen vector of the values. The current inplace ops for two sparse arena matrices is 

```cpp
  template <typename F, typename Expr,
            require_convertible_t<Expr&, MatrixType>* = nullptr,
            require_same_t<Expr, arena_matrix<MatrixType>>* = nullptr>
  inline void inplace_ops_impl(F&& f, Expr&& other) {
    auto&& x = to_ref(other);
    auto* val_ptr = (*this).valuePtr();
    auto* x_val_ptr = x.valuePtr();
    const auto non_zeros = (*this).nonZeros();
    for (Eigen::Index i = 0; i < non_zeros; ++i) {
      f(val_ptr[i], x_val_ptr[i]);
    }
  }
```

But we could instead for `+=` do 

```cpp
  template <typename F, typename Expr,
            require_convertible_t<Expr&, MatrixType>* = nullptr,
            require_same_t<Expr, arena_matrix<MatrixType>>* = nullptr>
  inline void operator+=(F&& f, Expr&& other) {
    auto&& x = to_ref(other);
    auto* val_ptr = (*this).valuePtr();
    auto* x_val_ptr = x.valuePtr();
    const auto non_zeros = (*this).nonZeros();
    Eigen::Map<MatrixType>(val_ptr, non_zeros) += Eigen::Map<MatrixType>(x_val_ptr, non_zeros);
  }
```

That would allow Eigen to optimize a lot more, but at the expense of losing a bit of generalization from just having `inplace_ops_impl`. If you'd like the above then I'm fine with the extra code imo